src/testing_tracer.rs

```diff
@@ -0,0 +1,191 @@
+    }
+}
+
+impl Tracer for TestingTracer {
+    type SupportedMemory = SimpleMemory;
+    const CALL_BEFORE_EXECUTION: bool = true;
+    const CALL_AFTER_DECODING: bool = true;
+
+    #[inline]
+    fn before_decoding(&mut self, _state: VmLocalStateData<'_>, _memory: &Self::SupportedMemory) {}
+
+    fn after_decoding(
+        &mut self,
+        _state: VmLocalStateData<'_>,
+        _data: AfterDecodingData,
+        _memory: &Self::SupportedMemory,
+    ) {
+        // check for built-in panics
+        if !_data.error_flags_accumulated.is_empty() {
+            // last accumulated panic will be used as exception_message
+            for (panic, _) in _data.error_flags_accumulated.iter_names() {
+                self.exception_message = Some(panic.to_owned());
+            }
+        }
+    }
+
+    fn before_execution(
+        &mut self,
+        _state: VmLocalStateData<'_>,
+        _data: BeforeExecutionData,
+        _memory: &Self::SupportedMemory,
+    ) {
+        let inner_opcode = _data.opcode.inner.variant.opcode;
+
+        // Propagate error message if Nop, ret.panic, ret.revert; reset otherwise
+        match inner_opcode {
+            Opcode::Nop(NopOpcode) => {}
+            Opcode::Ret(RetOpcode::Panic) => {}
+            Opcode::Ret(RetOpcode::Revert) => {}
+            _ => {
+                self.reset_exception();
+            }
+        }
+
+        // Try to execute commands
+        // commands always have r0 as src1 and dst0
+        if _data.opcode.src1_reg_idx == 0 && _data.opcode.dst0_reg_idx == 0 {
+            match inner_opcode {
+                Opcode::Add(AddOpcode::Add) => {
+                    // `add x r0 r0` is used as "execute x command" statement
+                    if self.expecting_register_value {
+                        self.execute_print_from_register(_data.src0_value);
+                    } else {
+                        if let (Some(command_prefix), Some(arg)) =
+                            self.parse_command_from_register(_data.src0_value)
+                        {
+                            if command_prefix == EXCEPTION_PREFIX {
+                                self.set_exception_message(&arg);
+                            } else if command_prefix == PRINT_PREFIX {
+                                self.execute_print(&arg);
+                            } else if command_prefix == PRINT_REG_PREFIX {
+                                self.expecting_register_value = true;
+                            } else if command_prefix == PRINT_PTR_PREFIX {
+                                self.expecting_register_value = true;
+                            }
+                        }
+                    }
+                }
+                Opcode::Ptr(PtrOpcode::Add) => {
+                    // `ptr.add x r0 r0` is used as "print" statement for pointers
+                    if self.expecting_register_value {
+                        self.execute_print_from_register(_data.src0_value);
```

Comment:
shouldn't we then set the 'expecing_register_value to 'false' then?

Reply:
Changed code structure to make it clearer